File: app/src/main/java/com/example/plantea/presentacion/fragmentos/CategoriasPictogramasFragment.java
package com.example.plantea.presentacion.fragmentos;

import android.app.Activity;
import android.content.Context;
import android.content.res.Configuration;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.Toast;

import com.example.plantea.R;
import com.example.plantea.dominio.Pictograma;
import com.example.plantea.presentacion.CrearPlanInterface;
import com.example.plantea.presentacion.adaptadores.AdaptadorPictogramas;

import java.util.ArrayList;


public class CategoriasPictogramasFragment extends Fragment implements AdaptadorPictogramas.OnItemSelectedListener {

    RecyclerView recyclerPictogramas;
    ArrayList<Pictograma> listaPictogramas;
    CrearPlanInterface interfaceCategorias;
    Activity actividad;
    View vista;

    private ImageView image_Cerrar;
    private ImageView image_add;

    public CategoriasPictogramasFragment() {
        // Required empty public constructor
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        vista = inflater.inflate(R.layout.fragment_categorias_pictogramas, container, false);

        Bundle bundle = this.getArguments();
        listaPictogramas = (ArrayList<Pictograma>) bundle.get("key");

        recyclerPictogramas =  vista.findViewById(R.id.recycler_Pictogramas);
        int orientation = getResources().getConfiguration().orientation;
        int gridValueManager;

        if (orientation == Configuration.ORIENTATION_PORTRAIT) {
            gridValueManager = 3; // set the number of columns to 2 for portrait mode
        } else {
            gridValueManager = 5; // set the number of columns to 3 for landscape mode
        }
        recyclerPictogramas.setLayoutManager(new GridLayoutManager(getContext(),gridValueManager ));
        image_Cerrar = vista.findViewById(R.id.image_Cerrar);
        image_add = vista.findViewById(R.id.image_add);

        AdaptadorPictogramas adaptador = new AdaptadorPictogramas(listaPictogramas,this);
        recyclerPictogramas.setAdapter(adaptador);

        //Este método se ejecutará al pinchar sobre la imagen de cerrar
        image_Cerrar.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                interfaceCategorias.cerrarFragment();
            }
        });

        //Este método se ejecutará al seleccionar añadir nuevo pictograma.
        image_add.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                interfaceCategorias.nuevoPictogramaDialogo();
            }
        });

        return vista;
    }


    @Override
    public void onAttach(@NonNull Context context) {
        super.onAttach(context);
        if(context instanceof Activity) {
            actividad = (Activity) context;
            interfaceCategorias = (CrearPlanInterface) actividad;
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
    }

    //Este método se ejecutará al seleccionar un pictograma para añadirlo en la planificacion
    @Override
    public void onItemSeleccionado(int posicion) {
        if (listaPictogramas.get(posicion).getCategoria() == 1){
            interfaceCategorias.mostrarsubCategoria(listaPictogramas.get(posicion).getTitulo());
        } else {
            interfaceCategorias.pictogramaSeleccionado(listaPictogramas.get(posicion).getTitulo(), listaPictogramas.get(posicion).getImagen(),listaPictogramas.get(posicion).getCategoria());
        }
    }
}

File: app/src/main/java/com/example/plantea/presentacion/fragmentos/CategoriasPictogramasFragment.kt
package com.example.plantea.presentacion.fragmentos

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.recyclerview.widget.RecyclerView
import com.example.plantea.R
import com.example.plantea.dominio.Pictograma
import com.example.plantea.presentacion.actividades.CommonUtils
import com.example.plantea.presentacion.adaptadores.AdaptadorPictogramas
import com.example.plantea.presentacion.viewModels.CrearPlanViewModel

class CategoriasPictogramasFragment : Fragment(), AdaptadorPictogramas.OnItemSelectedListener {
    lateinit var actividad: Activity

    lateinit var recyclerPictogramas: RecyclerView
    lateinit var textoVacio : TextView
    private lateinit var imageCerrar: ImageView
    private lateinit var imageAdd: Button
    private lateinit var constraintLayout: ConstraintLayout


    private val viewModel: CrearPlanViewModel by activityViewModels()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
        val vista = inflater.inflate(R.layout.fragment_categorias_pictogramas, container, false)

        recyclerPictogramas = vista.findViewById(R.id.recycler_Pictogramas)
        constraintLayout = vista.findViewById(R.id.frameLayout)

        context?.let { CommonUtils.getGridValueCuaderno(vista, context, recyclerPictogramas, constraintLayout, 150, 210) }

        imageCerrar = vista.findViewById(R.id.image_CerrarContenedor)
        imageAdd = vista.findViewById(R.id.image_add)
        textoVacio = vista.findViewById(R.id.textoNoPictogramas)
        val adaptador = AdaptadorPictogramas(viewModel._listaPictogramas.value, this, this)
        recyclerPictogramas.adapter = adaptador

        if(viewModel._listaPictogramas.value?.size == 0){
            textoVacio.visibility = View.VISIBLE
        }

        //Este método se ejecutará al pinchar sobre la imagen de cerrar
        imageCerrar.setOnClickListener {
            viewModel._closeFragment.value = true
        }

        if(viewModel.identificadorCategoria == 10){
            imageAdd.visibility = View.GONE
        }

        //Este método se ejecutará al seleccionar añadir nuevo pictograma.
        imageAdd.setOnClickListener { viewModel.nuevoPictogramaDialogo() }

        return vista
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is Activity) {
            actividad = context
        }
    }

    //Este método se ejecutará al seleccionar un pictograma para añadirlo en la planificacion
    override fun onItemSeleccionado(posicion: Int) {
       /* if (viewModel._listaPictogramas.value?.get(posicion)?.categoria in 1..4) {
            viewModel.mostrarsubCategoria(viewModel._listaPictogramas.value?.get(posicion)?.titulo, this.requireContext())
        } else {*/
            viewModel.pictogramaSeleccionado(posicion)
       // }
    }

    fun markAsFavorite(pictogram: Pictograma) {
        viewModel.pictograma.insertarFavorito(this.requireContext(), viewModel.idUsuario, pictogram.id, pictogram.titulo, pictogram.imagen, pictogram.sourceAPI)
    }

    fun removeFavorite(pictogram: Pictograma) {
        viewModel.pictograma.borrarFavorito(this.requireContext(), viewModel.idUsuario, pictogram.id)
    }
}


